Gems/PhysX/Code/Source/PhysXCharacters/Components/RagdollComponent.cpp

```diff
@@ -389,17 +406,62 @@ namespace PhysX
             }
         }
 
+        // If mass ratio clamping is enabled, iterate out from the root and clamp mass values
+        if (m_enableMassRatioClamping)
+        {
+            const float maxMassRatio = AZStd::GetMax(1.0f + AZ::Constants::FloatEpsilon, m_maxMassRatio);
+
+            // figure out the depth of each node in the tree, so that nodes can be visited from the root outwards
+            AZStd::vector<Utils::Characters::DepthData> nodeDepths =
+                Utils::Characters::ComputeHierarchyDepths(ragdollConfiguration.m_parentIndices);
+
+            AZStd::sort(
+                nodeDepths.begin(), nodeDepths.end(),
+                [](const Utils::Characters::DepthData& d1, const Utils::Characters::DepthData& d2)
+                {
+                    return d1.m_depth < d2.m_depth;
+                });
+
+            bool massesClamped = false;
+            for (const auto [depth, nodeIndex] : nodeDepths)
+            {
+                size_t parentIndex = ragdollConfiguration.m_parentIndices[nodeIndex];
+                if (parentIndex < numNodes)
+                {
+                    AzPhysics::RigidBody& nodeRigidBody = ragdoll->GetNode(nodeIndex)->GetRigidBody();
+                    const float originalMass = nodeRigidBody.GetMass();
+                    const float parentMass = ragdoll->GetNode(parentIndex)->GetRigidBody().GetMass();
+                    const float minMass = parentMass / maxMassRatio;
+                    const float maxMass = parentMass;
+                    if (originalMass < minMass || originalMass > maxMass)
+                    {
+                        const float clampedMass = AZStd::clamp(originalMass, minMass, maxMass);
+                        nodeRigidBody.SetMass(clampedMass);
+                        massesClamped = true;
+                        if (!AZ::IsCloseMag(originalMass, 0.0f))
```

Comment:
question: is this checking if `originalMass` is not close to 0.0f? What's the difference with `AZ::IsClose`?

Reply:
IsCloseMag is designed to take into account the size of the two things being compared, so for example it will use a different tolerance for comparing 1000 and 1001, vs comparing 0.001 and 0.002. But because I'm comparing to 0 here, IsClose is probably the more sensible choice, I'll update.

Reply:
Good to know about `IsCloseMag`, thanks for explaining.